Source/AMRInterpolator/SurfaceExtraction.hpp

```diff
@@ -0,0 +1,201 @@
+            return 2.0 * m_weights[weight_index];
+        else
+            return m_weights[weight_index];
+    }
+
+    static const IntegrationMethod trapezium;
+    static const IntegrationMethod midpoint;
+    static const IntegrationMethod simpson;
+    static const IntegrationMethod boole;
+};
+
+//! This class extracts grid variables on 2 dimensional surfaces each
+//! parameterised by u and v with different surfaces given by level sets of
+//! another parameter
+class SurfaceExtraction
+{
+  public:
+    struct params_t
+    {
+        int num_surfaces; //!< number of surfaces over which to extraction
+        std::vector<double>
+            surface_param_values; //!< the values of the
+                                  //!< parameter that gives the required
+                                  //!< surfaces with SurfaceGeom geometry (e.g.
+                                  //!< radii for spherical shells)
+        int num_points_u; //!< the number of points for the first parameter
+                          //!< that parameterises each surface
+        int num_points_v; //!< the number of points for the second parameter
+                          //!< that parameterises each surfaces
+        std::vector<int> extraction_levels; //!< the level on which to do the
+                                            //!< extraction for each surface
+        bool write_extraction; //!< whether or not to write the extracted data
+
+        int min_extraction_level()
+        {
+            return *(std::min_element(extraction_levels.begin(),
+                                      extraction_levels.end()));
+        }
+    };
+
+  protected:
+    const SurfaceGeometry *m_geom_ptr; //!< the pointer to the geometry class
+    const params_t m_params;
+    std::vector<std::pair<int, Derivative>> m_vars; //!< the vector of pairs of
+    //!< variables and derivatives to extract
+    const double m_dt;
+    const double m_time;
+    const bool m_first_step;
+    const double m_restart_time;
+    const int m_num_points; //!< the total number of points per surface
+    const double m_du;      //!< the grid spacing in u (used in integrate)
+    const double m_dv;      //!< the grid spacing in v (used in integrate)
+
+    std::vector<std::vector<double>> m_interp_data;
+    std::array<std::vector<double>, CH_SPACEDIM> m_interp_coords;
+
+    bool m_done_extraction; //!< whether or not the extract function has been
+                            //!< called for this object
+
+    //! returns the flattened index for m_interp_data and m_interp_coords
+    //! associated to given surface, u and v indices
+    int index(int a_isurface, int a_iu, int a_iv) const
+    {
+        return a_isurface * m_num_points + a_iu * m_params.num_points_v + a_iv;
+    }
+
+  public:
+    //! Normal constructor which requires vars to be added after construction
+    //! using add_var or add_vars
+    SurfaceExtraction(SurfaceGeometry *a_geom_ptr, params_t &a_params,
```

Comment:
If I can be really picky about this, SurfaceGeometry* and params_t can be const&, right? If yes, it also applies to the params_t in the other classes

Reply:
I agree with this mostly and have changed it but I don't see the point in having a const reference pointer so I just made it a const pointer.